src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/CpsProjectSystemReferenceReader.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.References;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Reference reader implementation for the core project system in the integrated development environment (IDE).
+    /// </summary>
+    internal class CpsProjectSystemReferenceReader
+        : IProjectSystemReferencesReader
+    {
+        private readonly IVsProjectAdapter _vsProjectAdapter;
+        private readonly IVsProjectThreadingService _threadingService;
+        private readonly AsyncLazy<UnconfiguredProject> _unconfiguredProject;
+
+        public CpsProjectSystemReferenceReader(
+            IVsProjectAdapter vsProjectAdapter,
+            IVsProjectThreadingService threadingService)
+        {
+            Assumes.Present(vsProjectAdapter);
+            Assumes.Present(threadingService);
+
+            _vsProjectAdapter = vsProjectAdapter;
+            _threadingService = threadingService;
+            _unconfiguredProject = new AsyncLazy<UnconfiguredProject>(async () =>
+            {
+                await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var context = _vsProjectAdapter.Project as IVsBrowseObjectContext;
+                if (context == null)
+                {
+                    // VC implements this on their DTE.Project.Object
+                    context = _vsProjectAdapter.Project.Object as IVsBrowseObjectContext;
+                }
+                return context?.UnconfiguredProject;
+            }, threadingService.JoinableTaskFactory);
+        }
+
+        public async Task<IEnumerable<ProjectRestoreReference>> GetProjectReferencesAsync(
+            Common.ILogger logger, CancellationToken _)
+        {
+            var unconfiguredProject = await _unconfiguredProject.GetValueAsync();
+            IBuildDependencyProjectReferencesService service = (await unconfiguredProject?.GetSuggestedConfiguredProjectAsync())?.Services.ProjectReferences;
```

Comment:
This will await a null and throw if UnconfiguredProject or GetSuggestedConfiguredProjectAsync returns null.